ir/src/main/scala/org/scalajs/core/ir/Infos.scala

```diff
@@ -64,4 +70,292 @@ object Infos {
+    private val accessedModules = mutable.Set.empty[String]
+    private val accessedClassData = mutable.Set.empty[String]
+
+    def setEncodedName(encodedName: String): this.type = {
+      this.encodedName = encodedName
+      this
+    }
+
+    def setIsStatic(isStatic: Boolean): this.type = {
+      this.isStatic = isStatic
+      this
+    }
+
+    def setIsAbstract(isAbstract: Boolean): this.type = {
+      this.isAbstract = isAbstract
+      this
+    }
+
+    def setIsExported(isExported: Boolean): this.type = {
+      this.isExported = isExported
+      this
+    }
+
+    def addMethodCalled(receiverTpe: Type, method: String): this.type = {
+      receiverTpe match {
+        case ClassType(cls) => addMethodCalled(cls, method)
+        case AnyType        => addMethodCalled(ObjectClass, method)
+        case UndefType      => addMethodCalled(BoxedUnitClass, method)
+        case BooleanType    => addMethodCalled(BoxedBooleanClass, method)
+        case IntType        => addMethodCalled(BoxedIntegerClass, method)
+        case LongType       => addMethodCalled(BoxedLongClass, method)
+        case FloatType      => addMethodCalled(BoxedFloatClass, method)
+        case DoubleType     => addMethodCalled(BoxedDoubleClass, method)
+        case StringType     => addMethodCalled(StringClass, method)
+
+        case ArrayType(_, _) if method != "clone__O" =>
+          /* clone__O is overridden in the pseudo Array classes and is
+           * always kept anyway, because it is in scalajsenv.js.
+           * Other methods delegate to Object, which we can model with
+           * a static call to Object.method.
+           */
+          addMethodCalledStatically(ObjectClass, method)
+
+        case _ =>
+          // Nothing to do
+      }
+
+      this
+    }
+
+    def addMethodCalled(cls: String, method: String): this.type = {
+      methodsCalled.getOrElseUpdate(cls, mutable.Set.empty) += method
+      this
+    }
+
+    def addMethodCalledStatically(cls: String, method: String): this.type = {
+      methodsCalledStatically.getOrElseUpdate(cls, mutable.Set.empty) += method
+      this
+    }
+
+    def addStaticMethodCalled(cls: String, method: String): this.type = {
+      staticMethodsCalled.getOrElseUpdate(cls, mutable.Set.empty) += method
+      this
+    }
+
+    def addInstantiatedClass(cls: String): this.type = {
+      instantiatedClasses += cls
+      this
+    }
+
+    def addInstantiatedClass(cls: String, ctor: String): this.type =
+      addInstantiatedClass(cls).addMethodCalledStatically(cls, ctor)
+
+    def addAccessedModule(cls: String): this.type = {
+      accessedModules += cls
+      this
+    }
+
+    def addAccessedClassData(tpe: ReferenceType): this.type = {
+      addAccessedClassData(tpe match {
+        case ClassType(name)    => name
+        case ArrayType(base, _) => base
+      })
+    }
+
+    def addAccessedClassData(cls: String): this.type = {
+      if (!AlwaysPresentClassData.contains(cls))
```

Comment:
This was also prompted by the fact that an `Is/AsInstanceOf(x, ClassType("jl_Integer"))` does not actually need the data of `jl_Integer`. This is the case for all hijacked classes. But I could also special-case the hijacked boxed classes in `Is/AsInstanceOf` and short-circuit them there. Maybe that's better. And then improve the compiler a tiny little bit to avoid generating `AsInstanceOf(x, ClassType("O"))`, which are no-ops, and are the main source of accessed-class-data "O".